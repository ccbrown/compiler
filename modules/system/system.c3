namespace system {

int64 write(int64 fd, char* buf, int64 len) {
	int64 ret;
	asm (
		"mov $$0x2000004, %rax\n"
		"mov  $1, %rdi    \n"
		"mov  $2, %rsi    \n"
		"mov  $3, %rdx    \n"
		"syscall\n"
		"mov %rax, $0\n"
		: "=m" (ret)
		: "m" (fd), "m" (buf), "m" (len)
		: "rax", "rdi", "rsi", "rdx"
	);
	return ret;
}

}