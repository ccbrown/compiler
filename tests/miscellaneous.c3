import system;

int64 write(int64 fd, void* buf, int64 len) { return 0; }

void main() {
	struct thing {
		int64 a;
		int64 b;
	}
	
	namespace foo {
		struct thing {
			int64 x;
		}
		
		int64 bar() {
			return 0;
		}
		
		thing lt;
		lt.x = 7;
	}
	
	foo::bar();
	
	foo::thing fb;
	
	fb.x = 3;
	
	int64 sum(thing* this) {
		return (*this).a + this->b;
	}
	
	thing t;
	
	t.a = 2;
	t.b = 3;
	
	int64 a = 50 / t.a;
	
	if (a == 25) {
		system::write(1, "pass\n", 5);
	}

	uint8 c = '0' + sum(&t);

	system::write(1, &c, 1);
	system::write(1, "\n", 1);
	
	int64 written = system::write(1, "hello world\n", 12);
	
	if (written == 1 + 3 * 2 + 5)
		system::write(1, "12\n", 3);
	
	if (1 == 2) {
		system::write(1, "??\n", 3);
		if (3 == 3) {
			return;
		}
	} else if (3 == 4 - 1) {
		system::write(1, "!!\n", 3);
	} else {
		system::write(1, "*1*\n", 4);
		system::write(1, "*2*\n", 4);
	}
	
	uint8* alloc = static_cast<uint8*>(system::calloc(1, 3));
	*alloc = 'x';
	*(alloc - -1) = '\n';
	system::write(1, alloc, 2);
	system::free(alloc);
	
	int64 i = 0;
	system::write(1, "loop\n", 5);
	while (i < 5) {
		uint8 x = '0' + i;
		system::write(1, &x, 1);
		system::write(1, "\n", 1);
		i = i + 1;
		
		if (i == 3) {
			return;
		}
	}

	return;
}