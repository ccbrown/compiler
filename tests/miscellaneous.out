pass
5
hello world
12
!!
x
loop
0
1
2
